src/libstd/alloc.rs

```diff
@@ -13,15 +13,77 @@
 #![unstable(issue = "32838", feature = "allocator_api")]
 
 #[doc(inline)] #[allow(deprecated)] pub use alloc_crate::alloc::Heap;
-#[doc(inline)] pub use alloc_crate::alloc::{Global, oom};
+#[doc(inline)] pub use alloc_crate::alloc::{Global, Layout, oom};
 #[doc(inline)] pub use alloc_system::System;
 #[doc(inline)] pub use core::alloc::*;
 
+use panicking::HOOK_LOCK;
+
+#[derive(Copy, Clone)]
+enum Hook {
+    Default,
+    Custom(*mut (Fn(Layout) -> ! + 'static + Sync + Send)),
+}
+
+static mut HOOK: Hook = Hook::Default;
+
+/// Registers a custom OOM hook, replacing any that was previously registered.
+///
+/// The OOM hook is invoked when an infallible memory allocation fails.
+/// The default hook prints a message to standard error and aborts the
+/// execution, but this behavior can be customized with the [`set_hook`]
+/// and [`take_hook`] functions.
+///
+/// The hook is provided with a `Layout` struct which contains information
+/// about the allocation that failed.
+///
+/// The OOM hook is a global resource.
+pub fn set_hook(hook: Box<Fn(Layout) -> ! + 'static + Sync + Send>) {
+    unsafe {
+        HOOK_LOCK.write();
+        let old_hook = HOOK;
+        HOOK = Hook::Custom(Box::into_raw(hook));
+        HOOK_LOCK.write_unlock();
+
+        if let Hook::Custom(ptr) = old_hook {
+            Box::from_raw(ptr);
+        }
+    }
+}
+
+/// Unregisters the current OOM hook, returning it.
+///
+/// *See also the function [`set_hook`].*
+///
+/// If no custom hook is registered, the default hook will be returned.
+pub fn take_hook() -> Box<Fn(Layout) -> ! + 'static + Sync + Send> {
+    unsafe {
+        HOOK_LOCK.write();
+        let hook = HOOK;
+        HOOK = Hook::Default;
+        HOOK_LOCK.write_unlock();
+
+        match hook {
+            Hook::Default => Box::new(default_hook),
+            Hook::Custom(ptr) => Box::from_raw(ptr),
+        }
+    }
+}
+
+fn default_hook(_: Layout) -> ! {
+    rtabort!("memory allocation failed")
```

Comment:
Might as well update this to something like `memory allocation of {} bytes failed`.